crates/wasmparser/src/readers/core/types.rs

```diff
@@ -165,20 +106,215 @@ impl<'a> FromReader<'a> for ValType {
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// Create a new `RefType`.
+    ///
+    /// Returns `None` when the heap type's type index (if any) is beyond this
+    /// crate's implementation limits and therfore is not representable.
+    pub fn new(nullable: bool, heap_type: HeapType) -> Option<Self> {
+        let nullable32 = if nullable { Self::NULLABLE_MASK } else { 0 };
+        match heap_type {
+            HeapType::TypedFunc(index) => RefType::typed_func(nullable, index),
+            HeapType::Func => Some(Self::from_u32(Self::ANY_FUNC_DISCRIMINANT | nullable32)),
+            HeapType::Extern => Some(Self::from_u32(Self::EXTERN_DISCRIMINANT | nullable32)),
+        }
+    }
+
+    const fn discriminant(&self) -> u32 {
+        self.as_u32() & Self::DISCRIMINANT_MASK
+    }
+
+    /// Is this a reference to a typed function?
+    pub const fn is_typed_func_ref(&self) -> bool {
+        self.discriminant() == Self::TYPED_FUNC_DISCRIMINANT
+    }
+
+    /// If this is a reference to a typed function, get its type index.
+    pub const fn type_index(&self) -> Option<u32> {
+        if self.is_typed_func_ref() {
+            Some(self.as_u32() & Self::INDEX_MASK)
+        } else {
+            None
+        }
+    }
+
+    /// Is this an untyped function reference aka `(ref null func)` aka `funcref` aka `anyfunc`?
+    pub fn is_func_ref(&self) -> bool {
+        self.discriminant() == Self::ANY_FUNC_DISCRIMINANT
+    }
+
+    /// Is this a `(ref null extern)` aka `externref`?
+    pub fn is_extern_ref(&self) -> bool {
+        self.discriminant() == Self::EXTERN_DISCRIMINANT
+    }
+
+    /// Is this ref type nullable?
+    pub const fn is_nullable(&self) -> bool {
+        self.as_u32() & Self::NULLABLE_MASK != 0
+    }
+
+    /// Get the non-nullable version of this ref type.
+    pub fn as_non_null(&self) -> Self {
+        Self::from_u32(self.as_u32() & !Self::NULLABLE_MASK)
+    }
+
+    /// Get the heap type that this is a reference to.
+    pub fn heap_type(&self) -> HeapType {
+        match self.discriminant() {
+            Self::TYPED_FUNC_DISCRIMINANT => HeapType::TypedFunc(self.type_index().unwrap()),
+            Self::ANY_FUNC_DISCRIMINANT => HeapType::Func,
+            Self::EXTERN_DISCRIMINANT => HeapType::Extern,
+            _ => unreachable!(),
+        }
+    }
+}
+
 impl<'a> FromReader<'a> for RefType {
     fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
         match reader.read()? {
             0x70 => Ok(RefType::FUNCREF),
             0x6F => Ok(RefType::EXTERNREF),
-            byte @ (0x6B | 0x6C) => Ok(RefType {
-                nullable: byte == 0x6C,
-                heap_type: reader.read()?,
-            }),
+            byte @ (0x6B | 0x6C) => {
+                let nullable = byte == 0x6C;
+                // NB: `FromReader for HeapType` only succeeds when
+                // `RefType::new` will succeed, so we can unwrap here.
```

Comment:
Could the condition be moved into this type perhaps through? Putting it in `HeapType` seems a bit odd when the restriction comes from `RefType` itself.